### PR TITLE
Update package to support new libp2p Transports/Upgraders

### PR DESCRIPTION
**Pull request by bigs**
This PR brings this package up to date with the latest APIs for go-libp2p. The transport, listener, and dialer behaviors have been updated to reflect the new interfaces and a potential information leak (via exposing onion addresses) has been shored up.

### CODE REVIEW

**Comment by bigs** [comment]
just occurred to me that it'd be useful to make a `NewOnionTransportC` function that returns a factory/constructor as well. going to add really quickly.

**Comment by bigs** [comment]
alright, that commit is up. this change should be good to go!

**Comment by cpacia** [comment]
Thanks @bigs I will give this a review

**Comment by bigs** [comment]
@Stebalien wise feedback. seems the listener and conn are worth remaining factored, but that's gone now.

File: onion_transport.go [hunk 1]
```diff
@@ -74,6 +76,62 @@ type OnionTransport struct {
...
+	protocols := []int{ma.P_ONION}
+	if !t.onlyOnion {
+		protocols = append(protocols, ma.P_TCP)
+	}
+	return protocols
+}
+
+// Proxy returns true, for tor is a relay-based protocol.
+func (t *OnionTransport) Proxy() bool {
+	return true
```
Comment (Stebalien) [inline onion_transport.go]: This should actually be false. Proxy is used when the rest of the multiaddr should be handled by some remote server.

This is an unfortunate consequence of how we dial. Ideally, we'd dial from left-to-right across the multiaddr, building up a connection from the ground up. However, this turned out to be really painful to implement (in go).

**Review by Stebalien — APPROVED** [review]

**Review by Stebalien — CHANGES_REQUESTED** [review]
(wrong button, should have been "request changes")

File: onion_transport.go (hunk 1 above)
Comment (bigs) [reply onion_transport.go]: ahhhhh, i see.

**Review by Stebalien — APPROVED** [review]

**Pull request closed without merging**